File: game-server/src/main/java/org/forfun/mmorpg/game/util/dirtywords/DirtyWordUnit.java
package org.forfun.mmorpg.game.util.dirtywords;

import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

@Getter
class DirtyWordUnit {

    private final String source;
    private final String keyWord;
    private final List<Integer> indexList = new ArrayList<>();

    public DirtyWordUnit(String source,String word) {
        this.source = source;
        this.keyWord = word;
    }

    public void checkWordIndex() {
        int index = 0;
        if (source.length() < keyWord.length()) {
            return ;
        }
        for (int i = 0,n=source.length(); i < n; i++) {
            if (keyWord.length() > index && source.charAt(i) == keyWord.charAt(index)) {
                indexList.add(index);
                index++;
                if (foundDirtyWord()) {
                    break;
                }
            }
        }
    }

    public boolean foundDirtyWord() {
        return this.indexList.size() == keyWord.length();
    }



}

File: game-server/src/main/java/org/forfun/mmorpg/game/util/dirtywords/DirtyWordUnit.kt
package org.forfun.mmorpg.game.util.dirtywords

import java.util.ArrayList

class DirtyWordUnit(private val source: String, word: String) {

    private val keyWord: String = word
    private val indexList: MutableList<Int> = ArrayList()

    fun checkWordIndex() {
        var index = 0
        if (source.length < keyWord.length) {
            return
        }
        var i = 0
        val n = source.length
        while (i < n) {
            if (keyWord.length > index && source[i] == keyWord[index]) {
                indexList.add(index)
                index++
                if (foundDirtyWord()) {
                    break
                }
            }
            i++
        }
    }

    fun foundDirtyWord(): Boolean {
        return this.indexList.size == keyWord.length
    }

    fun getSource(): String {
        return source
    }

    fun getKeyWord(): String {
        return keyWord
    }

    fun getIndexList(): List<Int> {
        return indexList
    }
}
